### ISSUE TITLE
Use a separate ECR-Repos for each environment

### ISSUE THREAD

**Issue by cmaerz** [aws/copilot-cli#4517]
We built a Pipeline that deploys a environment for each Merge Request we create on Gitlab. 

That works pretty well, if we wait about 3-4 Minutes between deploys, but if you start them more or less the same time there is a race condition and i can happen, that the wrong `latest` tag is deployed to the wrong enviroment.

Is there currently any way to prevent this? But I guess the best way is to use separate Repos for each env.

**Comment by paragbhingre** [comment]
Thank you for your request @cmaerz. I think this issue relates to https://github.com/aws/copilot-cli/issues/4045.

**Comment by cmaerz** [comment]
Yes it does. That killed me a production deployment last week, cause a co worker deployed a test environment the same time and the test environment got deployed to production. Tbh i guess that should be tackled pretty fast.

**Comment by Lou1415926** [comment]
Hi there @cmaerz ! Just curious, do you happen to run `copilot svc deploy --tag` ? Is the pipeline that you are using created via `copilot pipeline init`, or is it something else, e.g. GitLab CI/CD?

**Comment by cmaerz** [comment]
@Lou1415926 We use Gitlab CI/CD. I don't use `--tag` but if i use `--tag ${ENVIROMENT_NAME}` it tags the Enviroment Name instead of "latest"? Do i get that right?

**Comment by Lou1415926** [comment]
In fact, `latest` is always pushed. To be specific, as of Copilot v1.26.0:
1. If `--tag mytag` is specified, then the image is tagged with `latest` and `mytag`. Copilot passes `imageURL:mytag` to CloudFormation to reference the image.
2. Otherwise, if the workspace is a git repo and does not have uncommitted change, the image is tagged with `latest` and the `3ec52` (the abbreviated commit hash). Copilot then passes `imageURL:3ec52` to CloudFormation.
3. Otherwise, the image is tagged with `latest` only. If Copilot got a image digest from `docker push`, then it passes `imageURL@sha256:b3c3bdb6fdsf3346sds` to CloudFormation, the part after `@` being the digest. Otherwise, it falls back to passing `imageURL:latest`.


If I'm not mistaken, it seems that any of `imageURL:mytag`, `imageURL:3ec52` and `imageURL@sha256:b3c3bdb6fdsf3346sds` is sufficient to distinguish your test image and prod image. It's the last case that could cause issue for you. I am not sure why `docker push` didn't return Copilot an image digest though 🤔 

Just to double check, the value to the `ContainerImage` parameter in your CloudFormation template is indeed `imageURL:latest` right?

**Comment by cmaerz** [comment]
Ok thats funny. I guess I know what happened now:
1. First I pushed to my `main` branch to deploy to our staging env
2. I created a Branch from main w/o a new commit and created a Merge Request (needed to test if my automated Test Environments work)
2.1. This triggers another Pipelines that creates a new ENV but somehow it was faster than the staging Pipeline and the staging Pipeline pulled the image created by the Merge Request on the same Commit and deployed it to the staging env.

It's a small Edge case and i guess you can prevent it pretty easy with `--tag ${ENV}`

**Comment by Lou1415926** [comment]
@cmaerz yeah in this case it sounds like we can give `--tag ${Env}` or something like `--tag ${Env}-<commit-hash>` a try 👍🏼 

I do wonder if Copilot should have prioritized referencing the image by image digest before it decides to use the commit hash:
1. If `--tag mtag` is specified, tag `latest` and `mytag`; uses `imageURL:mytag`
2. Otherwise, if there is a clean commit, tag `latest` and `3ec52`, but use `imageURL@sha256: b3c3bdb6fdsf3346sds` when possible, and fallback to `imageURL:3ec52`
3. Tag with `latest` only; use `imageURL@sha256: b3c3bdb6fdsf3346sds` when possible, and fallback to `imageURL:latest`

Leaving this thought here for us to consider!

**Comment by cmaerz** [comment]
I use --tag now. Works fine for me :)
